[Logstash] Add CEL native data collection, and dashboards for Pipeline and Pipeline Details

robbavey opened the pull request:
## Proposed commit message

This PR follows on from #7704, and adds data collection to the integration, as well as standard Dashboards for Logstash Pipelines.

The short term goal for this is to enable Logstash to be monitored with Serverless Elasticsearch instances, which do not provide a Stack Monitoring UI, with the long term goal of iterating on this to replace the existing Stack Monitoring UI.

This PR includes 3 new data streams - `node_cel`, `pipelines` and `plugins`, with data collected from the Logstash Metrics API using the CEL input, intended to replace the existing metricbeat collection. 

This give us more flexibility going forward, and should improve our velocity to add new features to the Logstash dashboards straight from the API, without needing to update several code bases.

The CEL data streams will create a single document per node for the `node_cel` data stream, a document per pipeline per node for the `pipeline` data stream, and a document per plugin per pipeline per node for the `plugins` data stream. For this reason, we default to a lower sending cadence for the `plugins` data stream than for the other. 

Note that the existing data collection from metricbeat defaults to sending large documents, including significant data for the entire pipeline graph every 10 seconds. 

This integration also includes ingest pipelines to clean up the data from the Logstash flow metrics API, to remove entries we don't necessarily need at this stage, and to remove the string representation of "Infinity".

When installing a Logstash integration, users will be presented with an option to collect Metrics for Stack Monitoring or for Technical Preview

<img width="698" alt="Screenshot 2023-10-04 at 3 50 30 PM" src="https://github.com/elastic/integrations/assets/4061337/15e92e5e-c4fb-4813-862f-b1cd655fc4ce">

(Note: If there was a way to have an either/or option in the UI, that would be very helpful)

The dashboards are described in the _Screenshots_ section.



## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).


## How to test this PR locally


* Install [elastic-package](https://github.com/elastic/elastic-package)
* Checkout this PR
* Navigate to the `integrations/packages/logstash` directory
* Run `elastic-package build` to build the package
* Run `elastic-package stack up` to bring up the Elastic stack
* Start an instance of Logstash
* Install the `Logstash` integration
* Uncheck the `Metrics (Stack Monitoring)` option
* Check the `Metrics (Technical Preview)` option
* Click 'Save and Continue' and follow the instructions to enroll the agent

## Related issues

Follows on from #7704

- 

## Screenshots

Integration Selection Page


<img width="812" alt="Screenshot 2023-10-04 at 3 56 22 PM" src="https://github.com/elastic/integrations/assets/4061337/f304f47b-b63f-4b52-b5cf-efc302117bc3">


### Pipelines Dashboard

This Dashboard provides an overview of running pipelines, with a table containing details of pipelines running, which can be drilled down into a "details view"

#### Proposed Page

<img width="1770" alt="Screenshot 2023-10-04 at 4 08 37 PM" src="https://github.com/elastic/integrations/assets/4061337/0195d157-fc20-4f2f-8f23-2cb2a1c4e11f">


This is intended to replace the existing pipelines page in the Stack monitoring solution

##### Existing Page
<img width="1765" alt="Screenshot 2023-10-04 at 4 08 15 PM" src="https://github.com/elastic/integrations/assets/4061337/0ab3f36b-fb27-4a04-b1d4-a9d4872b1fb7">


The page includes additional details, including Persistent Queue usage per pipeline, and more visibility into which pipelines may be blocked


### Pipeline Details Dashboard

The Pipeline Details dashboard takes a different approach to the pipeline details view in stack monitoring

The page includes some general pipeline information, showing the rate of events flowing through the pipelines and state of persistent queues, broken down per node running the pipeline:

#### Proposed Page

##### General Information

<img width="1775" alt="Screenshot 2023-10-04 at 4 26 05 PM" src="https://github.com/elastic/integrations/assets/4061337/4675b0f1-9652-4535-bd42-874a3d45630c">

It also includes information for each plugin type: 

##### Inputs

<img width="1478" alt="Screenshot 2023-10-04 at 4 27 53 PM" src="https://github.com/elastic/integrations/assets/4061337/a3bcbfee-27c0-4ee6-813a-2598961f97cb">

##### Filters

<img width="1480" alt="Screenshot 2023-10-04 at 4 28 50 PM" src="https://github.com/elastic/integrations/assets/4061337/ba501884-3b2d-4855-a200-c2aa5f61b4c9">

###### Outputs

<img width="1478" alt="Screenshot 2023-10-04 at 4 29 49 PM" src="https://github.com/elastic/integrations/assets/4061337/2af4fe5f-ce7e-4cdd-b07c-79db1ea74f5f">

###### Codecs

<img width="1477" alt="Screenshot 2023-10-04 at 4 30 21 PM" src="https://github.com/elastic/integrations/assets/4061337/f2cf75d0-26ff-4f5d-b16a-3fbba086be4e">

This has benefits and drawbacks compared with the existing approach:

The benefits are:
* More information and more rich drill downs to help track down slow performing plugins
* Can track PQ usage on a per-pipeline and per-node basis

Drawbacks:
* The lack of visual representation may make it more difficult to figure out which plugin is at fault when multiple plugins of the same type are at play
  * Note - this can be mitigated by users including an id in their plugin definition
  * We will be looking to add more information (originating file + line number of plugin) in a subsequent follow-up PR.
 
#### Existing Dashboard

The existing dashboard contains a representation of the pipeline graph, which we are not easily able to show using standard Kibana Dashboards

<img width="346" alt="Screenshot 2023-10-04 at 4 32 50 PM" src="https://github.com/elastic/integrations/assets/4061337/b067d37a-9449-415d-af8e-9c29879a200b">

And a small pop-up drilldown for each plugin

<img width="439" alt="Screenshot 2023-10-04 at 4 36 56 PM" src="https://github.com/elastic/integrations/assets/4061337/0b683ac4-3e6c-4d35-8944-9402ac562c90">



Note: This will require a follow-up PR to enable for serverless - the existing integration has some compatibility issues that I would like to fold into a subsequent PR

elasticmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-8098/detail/PR-8098/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-8098/detail/PR-8098/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-8098/detail/PR-8098/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-8098/detail/PR-8098/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://integrations_8098.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/fleet-ci/transactions/view?rangeFrom=2023-10-04T20:49:43.631Z&rangeTo=2023-10-04T21:09:43.631Z&transactionName=BUILD+Ingest-manager%2Fintegrations%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=f69125eab7ddf41610f11d3574917cc0&transactionId=0732373c8f77281b)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2023-10-04T20:59:43.631+0000


* Duration: 15 min 43 sec





</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->














### Steps errors [![3](https://img.shields.io/badge/3%20-red)](https://fleet-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-8098/detail/PR-8098/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `Test integration: logstash`
<ul>
<li>Took 0 min 19 sec . View more details <a href="https://fleet-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Ingest-manager/pipelines/integrations/pipelines/PR-8098/runs/1/steps/1209/log/?start=0">here</a></li>
<li>Description: <code>                                eval "$(../../build/elastic-package stack shellinit)"                                ../../build/elastic-package test -v --report-format xUnit --report-output file --test-coverage                                </code></l1>
</ul>
  
  
  
  
##### `Google Storage Download`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://fleet-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Ingest-manager/pipelines/integrations/pipelines/PR-8098/runs/1/steps/1298/log/?start=0">here</a></li>

</ul>
  
  
  
  
##### `Google Storage Download`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://fleet-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Ingest-manager/pipelines/integrations/pipelines/PR-8098/runs/1/steps/1311/log/?start=0">here</a></li>

</ul>
  
  </p>
  </details>



## :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.




</p>
</details>

<!--COMMENT_GENERATED_WITH_ID_comment.id-->

robbavey commented:
/test

elasticmachine commented:
### :globe_with_meridians: Coverage report

Name | Metrics % (`covered/total`) | Diff
 --- | --- | ---
Packages | 100.0% (`2/2`) | :green_heart: 
Files | 100.0% (`6/6`) | :green_heart: 3.659
Classes | 100.0% (`6/6`) | :green_heart: 3.659
Methods | 100.0% (`55/55`) | :green_heart: 13.301
Lines | 87.558% (`190/217`) | :-1: -4.497
Conditionals | 100.0% (`0/0`) | :green_heart: 

<!--COMMENT_GENERATED_WITH_ID_coverage-->

andrewkroh commented on packages/logstash/data_stream/node_cel/agent/stream/stream.yml.hbs:
My personal preference is to name this file based on the input type used like `cel.yml.hbs`. If you do so, then update the `template_path` attribute in the manifest.yml.

andrewkroh commented on packages/logstash/data_stream/node_cel/agent/stream/stream.yml.hbs:
```diff
@@ -0,0 +1,71 @@
+  password: {{password}}
+
+redact:
+  fields: ~
```
```suggestion
  fields:
    - password
```

andrewkroh replied:
But if you switch to using `auth.basic` then you won't need to redact it or configure the request with the Authorization header manually.

```
auth:
  basic:
    user: {{escape_string username}}
    password: {{escape_string password}}

andrewkroh commented on packages/logstash/validation.yml:
```diff
@@ -0,0 +1,3 @@
+errors:
+  exclude_checks:
+    - SVR00002  # Mandatory filters in dashboards.
```
I think this needs an `elastic-package format`.

andrewkroh commented on packages/logstash/data_stream/node_cel/fields/fields.yml:
```diff
@@ -0,0 +1,270 @@
+                    - name: cpuacct
+                      type: group
+                      fields:
+                        - name: control_group
```
There's a little inconsistency between node_cel and node_stats. `control_group` is indexed as `text` in at packages/logstash/data_stream/node_stats/fields/fields.yml:128:27.  Same for the *.cpu.control_group.

That might cause some query/agg issues.

andrewkroh commented on packages/logstash/data_stream/node_cel/fields/agent.yml:
```diff
@@ -0,0 +1,141 @@
+      type: keyword
+      ignore_above: 1024
+      description: Host mac addresses.
+    - name: name
```
This field is already declared in ecs.yml.

andrewkroh commented on packages/logstash/data_stream/node_cel/fields/agent.yml:
```diff
@@ -0,0 +1,141 @@
+  footnote: "Examples: If Metricbeat is running on an EC2 host and fetches data from its host, the cloud info contains the data about this machine. If Metricbeat runs on a remote machine outside the cloud and fetches data from a service running in the cloud, the field contains cloud data from the machine the service is running on."
+  type: group
+  fields:
+    - name: account.id
+      type: keyword
+      ignore_above: 1024
+      description: "The cloud account or organization id used to identify different entities in a multi-tenant environment.\nExamples: AWS account id, Google Cloud ORG Id, or other unique identifier."
```
A lot of these fields exist in ECS and should use `external: ecs`.

```suggestion
    - name: account.id
      external: ecs
```

I've been working on a tool to fix this automatically. This command will edit the file for you. (I would only use it on a clean workspace so you can see review its edits.)

`go run github.com/andrewkroh/fydler@main -fix packages/logstash/data_stream/node_cel/fields/*yml`

efd6 commented on packages/logstash/data_stream/node_cel/agent/stream/stream.yml.hbs:
```diff
@@ -0,0 +1,71 @@
+
+program: |
+  (has(state.username) && state.username != null ?
+    get_request(state.url).with({
+        "Header":{
+            "Authorization": ["Basic "+string(base64(state.username+":"+state.password))],
+        }
+    })
```
This can be done more easily withe the [`basic_authentication` extension](https://pkg.go.dev/github.com/elastic/mito/lib#hdr-Basic_Authentication).

robbavey replied:
Is there something I am missing beyond adding `auth.basic.user` and `auth.basic.password` blocks eg - 

```
config_version: "2"
interval: {{period}}
resource.url: "{{url}}/_node/stats?graph=true&vertices=true"
{{#if resource_ssl}}
resource.ssl:
  {{resource_ssl}}
{{/if}}

{{#if username}}
auth.basic.user: {{escape_string username}}
{{/if}}
{{#if password}}
auth.basic.password: {{escape_string password}}
{{/if}}

redact:
  fields: ~

program: |
  get_request(state.url)
  .do_request().as(resp, bytes(resp.Body)
:
:
```

I have tried this a number of times, and I've never seen the `Authorization` header sent successfully, hence the workarounds applied here.

efd6 commented on packages/logstash/data_stream/node_cel/agent/stream/stream.yml.hbs:
```diff
@@ -0,0 +1,71 @@
+  .decode_json()
+  .as(body,
+    {"logstash":{"node":{"stats":{
+             "events":body['events'],
+             "jvm":{
+                "uptime_in_millis":body['jvm']['uptime_in_millis'],
```
These and below can be dotted, eg:
```suggestion
             "events":body.events,
             "jvm":{
                "uptime_in_millis":body.jvm.uptime_in_millis,
```
Stylistic only, but slightly lighter on the page.

efd6 commented on packages/logstash/data_stream/node_cel/agent/stream/stream.yml.hbs:
```diff
@@ -0,0 +1,71 @@
+        }})
+  )
+  .as(eve, {
+    "events":eve.map(each, eve),
```
What is the intention here? It looks to me like you are wanting a single element array with `eve` as the element. Is that correct?

```suggestion
    "events": [eve],
```

robbavey replied:
In this case, yes - your suggestion works perfectly

efd6 commented on packages/logstash/data_stream/plugins/agent/stream/stream.yml.hbs:
```diff
@@ -0,0 +1,174 @@
+  .do_request().as(resp, bytes(resp.Body)
+  .decode_json().as(body,
+    body.pipelines.map(pipeline_name, pipeline_name != ".monitoring-logstash", {"name":pipeline_name}.with(body.pipelines[pipeline_name])
+      .with({
+        "es_cluster_id":((body.pipelines[pipeline_name].vertices).as(vertices, vertices.map(each, has(each.cluster_uuid), each.cluster_uuid)))
+      })
+      .with({
+        "es_cluster_id_map":((body.pipelines[pipeline_name].vertices).as(vertices, vertices.map(each, has(each.cluster_uuid), {"plugin_id":each.id, "cluster_id":each.cluster_uuid})))
+      })
+      .with({
+        "outputs":body.pipelines[pipeline_name]["plugins"]["outputs"]
+      })
+      .with({
+        "inputs":body.pipelines[pipeline_name]["plugins"]["inputs"]
+      })
+      .with({
+        "filters":body.pipelines[pipeline_name]["plugins"]["filters"]
+      })
+      .with({
+        "codecs":body.pipelines[pipeline_name]["plugins"]["codecs"]
+      })
+      .with({
+        "host":{
+          "name":body.name,
+          "address":body.http_address,
+        }
+      })
```
Each `with` call will need to allocate a Go map. I would combine all these into one.
```suggestion
      .with({
        "es_cluster_id":((body.pipelines[pipeline_name].vertices).as(vertices, vertices.map(each, has(each.cluster_uuid), each.cluster_uuid))),
        "es_cluster_id_map":((body.pipelines[pipeline_name].vertices).as(vertices, vertices.map(each, has(each.cluster_uuid), {"plugin_id":each.id, "cluster_id":each.cluster_uuid}))),
        "outputs":body.pipelines[pipeline_name].plugins.outputs,
        "inputs":body.pipelines[pipeline_name].plugins.inputs,
        "filters":body.pipelines[pipeline_name].plugins.filters,
        "codecs":body.pipelines[pipeline_name].plugins.codecs,
        "host":{
          "name":body.name,
          "address":body.http_address,
        }
      })
```
Maybe also consider pulling out `body.pipelines[pipeline_name].plugins` once, before the evaluation here.

(oops, missed the initial `with`, but you get the idea)

efd6 commented on packages/logstash/data_stream/plugins/agent/stream/stream.yml.hbs:
```diff
@@ -0,0 +1,174 @@
+                "ms":input.events.queue_push_duration_in_millis
+              }
+            }
+          }.drop_empty()
```
Check to see that you still get what you want with only the final `drop_empty` (L87) in the expression.

efd6 commented on packages/logstash/data_stream/plugins/agent/stream/stream.yml.hbs:
```diff
@@ -0,0 +1,174 @@
+                "ms":filter.events.duration_in_millis
+              }
+            }
+          }.drop_empty()
```
Similar here.

efd6 commented on packages/logstash/data_stream/plugins/agent/stream/stream.yml.hbs:
```diff
@@ -0,0 +1,174 @@
+      "cursor": {"username":(has(state.cursor) && has(state.cursor.username)) ? state.cursor.username : has(state.username) ? state.username : null,
+                 "password":has(state.cursor) && has(state.cursor.password) ? state.cursor.password : has(state.password) ? state.password : null
+                 },
+      "events":eve.map(each, {"logstash":{"pipeline":each}})})
```
```suggestion
      "events":[{"logstash":{"pipeline":each}})}]
```

robbavey replied:
This doesn't work, as far as I can tell - I am trying to embed each of the events produced earlier in the script, into an array of `{"logstash":{"pipeline":{BODY}}}` events, so I am using the `map` function to do that.

efd6 replied:
You are correct.

robbavey commented:
I think I've addressed most of the comments in the original reviews - apart from the simplified `basic_auth`, which I have been unable to get working thus far

efd6 commented on packages/logstash/data_stream/pipeline/agent/stream/cel.yml.hbs:
```diff
@@ -0,0 +1,72 @@
+      .with({
+        "elasticsearch.cluster.id":((body.pipelines[pipeline_name].vertices).as(vertices, vertices.map(each, has(each.cluster_uuid), each.cluster_uuid)))
+      })
+      .with({
+        "host":{
+          "name":body.name,
+          "address":body.http_address,
+        }
+      })
+      .with({
```
You can merge these as well.

efd6 commented on packages/logstash/data_stream/pipeline/agent/stream/cel.yml.hbs:
```diff
@@ -0,0 +1,72 @@
+          "flow":body.pipelines[pipeline_name].flow,
+          "time":{
+            "queue_push_duration":{
+              "ms":body.pipelines[pipeline_name]["events"]["queue_push_duration_in_millis"],
```
```suggestion
              "ms":body.pipelines[pipeline_name].events.queue_push_duration_in_millis,
```
(similar below — quoted bracket indexing is only needed when the key is not a valid CEL token)

efd6 commented on packages/logstash/data_stream/pipeline/agent/stream/cel.yml.hbs:
```diff
@@ -0,0 +1,72 @@
+{{/if}}
+
+redact:
+  fields:
+    - password
```
This can now be the null case.
```suggestion
  fields: ~
```

efd6 commented on packages/logstash/data_stream/plugins/agent/stream/cel.yml.hbs:
```diff
@@ -0,0 +1,150 @@
+{{/if}}
+
+redact:
+  fields:
+    - password
```
```suggestion
  fields: ~
```

efd6 commented on packages/logstash/data_stream/plugins/agent/stream/cel.yml.hbs:
```diff
@@ -0,0 +1,150 @@
+      .with({
+        "es_cluster_id":((body.pipelines[pipeline_name].vertices).as(vertices, vertices.map(each, has(each.cluster_uuid), each.cluster_uuid))),
+        "es_cluster_id_map":((body.pipelines[pipeline_name].vertices).as(vertices, vertices.map(each, has(each.cluster_uuid), {"plugin_id":each.id, "cluster_id":each.cluster_uuid}))),
+        "outputs":body.pipelines[pipeline_name]["plugins"]["outputs"],
```
```suggestion
        "outputs":body.pipelines[pipeline_name].plugins.outputs,
```
(similar below)

efd6 commented on packages/logstash/data_stream/plugins/agent/stream/cel.yml.hbs:
```diff
@@ -0,0 +1,150 @@
+        "plugin":{
+          "type":"input",
+          "input":{
+            "elasticsearch.cluster.id":event['es_cluster_id_map'].map(tuple, (tuple.plugin_id == input.id), tuple.cluster_id),
```
```suggestion
            "elasticsearch.cluster.id":event.es_cluster_id_map.map(tuple, (tuple.plugin_id == input.id), tuple.cluster_id),
```
(similar below)

robbavey commented:
This is ready for another round

robbavey commented on packages/logstash/data_stream/pipeline/agent/stream/cel.yml.hbs:
```diff
@@ -0,0 +1,65 @@
+          },
+          "events":{
+            "out":body.pipelines[pipeline_name].events.out,
+            "in":body.pipelines[pipeline_name].events["in"],
+            "filtered":body.pipelines[pipeline_name].events.filtered,
+          },
+          "queues":{
+            "type":body.pipelines[pipeline_name].queue["type"],
```
I've made the change to `queue.type`, but I can't make the change to `events.in` - it fails to compile with 

```
Syntax error: no viable alternative at input '.in'
 |           "in":body.pipelines[pipeline_name].events.in,
```

Which was the reason I changed to square brackets from dotted in the first place...

efd6 replied:
OK. That's the criterion I use, trust the compiler over me.

robbavey commented:
This is ready for another round @elastic/infra-monitoring-ui 

Note, that this is part of the serverless initiative to enable Logstash monitoring in the absence of a stack monitoring UI

I will be following on this PR with the changes necessary to make this package visible to serverless, which includes changes to the existing integration package

klacabane reviewed (approved):
LGTM!

IanLee1521 commented:
First off, I just found this and it's AWESOME!

Maybe it's better for me to open a new issue, but just before I do, would it make sense to mark these dashboards with a reference to users needing to switch the integration from Stack Monitoring to Technical Preview?

I just ran in to this where I was trying to look at the `[Metrics Logstash] *` dashboards and getting errors. 

![image](https://github.com/elastic/integrations/assets/828452/49ccbd62-1f89-47ea-8a71-b911709ec13c)

Eventually I found this need to change the integration configuration and now things are working.

Thoughts? Is that worth noting on the dashboards?

robbavey commented:
@IanLee1521 First of all, thanks for trying out the new dashboards in technical preview - we love to get feedback on new features we are developing, and we appreciate you trying them out as we refine them. Any further comments would be gratefully received.

I think this should be an issue - as we start to transition to the new way of collecting and presenting data for monitoring, we should be making it clearer how to work with this effectively.

I'm on PTO at the moment, but I can make an issue when I get back, unless you would prefer to create one yourself.

robbavey commented:
@IanLee1521 - I've added an issue [here](https://github.com/elastic/integrations/issues/8643). And thanks again for the feedback!
